### ISSUE TITLE
[🐛 webdriverio cant connect to webdrivers for chrome and edge

### ISSUE THREAD

**Issue by rpii** [webdriverio/webdriverio#12188]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

8.29.1 thru 8.29.7

### Node.js Version

20.4.0

### Mode

Standalone Mode

### Which capabilities are you using?

```typescript
capabilities: [
        {
            // Set maxInstances to 1 if screen recordings are enabled:
            maxInstances: 1,
            browserName: 'chrome',
            'goog:chromeOptions': {
                args: [
                    "--acceptInsecureCerts=true", 
                    "--ignore-ssl-errors",
                    "--window-size=1600,1400",
                    "--ignore-certificate-errors",
                    ]
            },
            'wdio:chromedriverOptions':  {
                binary: currentDirectory + "/webdrivers/" + binary,
                logLevel: "ALL",
                enableChromeLogs:true,
                logPath:"logs"
            } 
        }
    ]
```

### What happened?

run yarn test:chrome
the webdriver starts but there is a connect error when sending the session message to the driver. happens in chrome and edge

### What is your expected behavior?

starts the webdriver and runs the test.

### How to reproduce the bug.

to reproduce  pull code from git@github.com:rpii/wdio-webdriver-test.git
run directly yarn test:chrome or yarn test:edge.  

### Relevant log output

```typescript
Starting ChromeDriver 121.0.6167.85 (3f98d690ad7e59242ef110144c757b2ac4eef1a2-refs/branch-heads/6167@{#1539}) on port 57026
Remote connections are allowed by an allowlist (0.0.0.0).
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.

2024-02-06T22:41:27.230Z INFO @wdio/local-runner: Run worker command: run
2024-02-06T22:41:27.334Z DEBUG @wdio/runner: init remote session
2024-02-06T22:41:27.340Z DEBUG @wdio/utils:shim: Finished to run "beforeSession" hook in 0ms
2024-02-06T22:41:27.471Z DEBUG @wdio/runner: init remote session
2024-02-06T22:41:27.471Z INFO webdriver: Initiate new session using the WebDriver protocol


error [ERR_STREAM_WRITE_AFTER_END]: write after end
[0-0]     at new NodeError (node:internal/errors:400:5)
[0-0]     at _write (node:internal/streams/writable:322:11)
[0-0]     at WriteStream.Writable.write (node:internal/streams/writable:337:10)
[0-0]     at Socket.ondata (node:internal/streams/readable:766:22)
[0-0]     at Socket.emit (node:events:513:28)
[0-0]     at Socket.emit (node:domain:489:12)
[0-0]     at addChunk (node:internal/streams/readable:324:12)
[0-0]     at readableAddChunk (node:internal/streams/readable:297:9)
[0-0]     at Socket.Readable.push (node:internal/streams/readable:234:10)
[0-0]     at Pipe.onStreamRead (node:internal/stream_base_commons:190:23) {
[0-0]   code: 'ERR_STREAM_WRITE_AFTER_END'
[0-0] }
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by christian-bromann** [comment]
I can't run the reproducible example on Mac:

```

node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
Error: Qualified path resolution failed: we looked for the following paths, but none could be accessed.

Source path: /private/tmp/wdio-webdriver-test/node_modules/@wdio/cli/bin/wdio.js
Not found: /private/tmp/wdio-webdriver-test/node_modules/@wdio/cli/bin/wdio.js
```

Any reasons why you run your custom downloaded driver?

**Comment by rpii** [comment]
Yes it is a requirement you might consider.  we run all of our tests on an air gapped network.  We have no connectivity to the internet.   all files are scanned and transferred  manually.

**Comment by rpii** [comment]
we do not support macs nor have I ever tried to run anything there.  I tested this on two windows 11 boxes

**Comment by christian-bromann** [comment]
It seems like something fails when trying to capture the stdout/stderr from the driver process based on given driver capabilities. Any investigations or contributions would be appreciated.

**Comment by rpii** [comment]
Well if I knew how to debug it effectively I would have done that.  I turned all the logging I could.   as you see there is no other output.  I have seen a crash when a driver tries to redirect to the log file,  that was hundreds of runs ago.  I think chromedriver crashes if the specified log folder doesnt exist

**Comment by rpii** [comment]
Is there any good tutorial on debugging webdriverio internals?  I have
tried in the past and the experience was like looking for a needle in a
haystack.   I would be glad to work on this if I could be useful.

thanks
Rich

**Comment by christian-bromann** [comment]
> Is there any good tutorial on debugging webdriverio internals?

I would directly put in some `console.log(...)` in the WebdriverIO modules within your `node_modules`. You want to look into the `@wdio/utils` package where we handle drivers and browsers.

**Comment by rpii** [comment]
So I walked through all of the options and found the bad ones:
`    capabilities: [
        {
            // Set maxInstances to 1 if screen recordings are enabled:
            maxInstances: 1,
            browserName: 'chrome',
            'goog:chromeOptions': {
                args: [
                    "--ignore-ssl-errors",
                    "--window-size=1600,1200",
                    "--ignore-certificate-errors",
                    ]
            },
            'wdio:chromedriverOptions':  {
                binary: currentDirectory + "/webdrivers/" + binary,
                logLevel: "ALL",
                // enableChromeLogs:true,
                // logPath:"logs"
            } 
        }
    ]`

logPath: "logs" is poison.  brings everything down.
enableChromeLogs:true    it causes an extra console window to appear with each browser load.  sort of runs though.

so with the above config I can run.  

No joy for edge. I can not get that to run.  even the wdio cli project wont bring up  a browser.  I will look further at that.

**Comment by rpii** [comment]
Further investigation with the edge driver shows that the two options 
`
                enableChromeLogs:true,
                logPath:"logs"
`

prevent it from working as well.

**Comment by erwinheitzman** [comment]
@rpii nice finds, do you feel like any actions are still needed or can we close this?

**Comment by rpii** [comment]
While there are defects in the logging settings, they can be worked around.
Edge however does not work in Linux. I have not been able to find any
settings that will work

**Comment by christian-bromann** [comment]
I think it is worth checking if any of the logging settings prevent a WebdriverIO session to start. If so we should ensure that these settings are respected while a session can always be created.

**Comment by rpii** [comment]
I documented that in a previous email. I can repost it if need be

**Comment by RafalSkorka** [comment]
I am also experiencing `ERR_STREAM_WRITE_AFTER_END` when the `logPath` option is used. I really would like to use it because without it, I cannot use `readableTimestamp` option (see [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/test/chromedriver/logging.cc;l=284?q=readable-timestamp&ss=chromium) for the relevant if statement).
I run two parallel suites, each has ~45 tests and this problem happens ~8 times during the execution.

**Comment by Delta456** [comment]
Is this issue still relevant?

**Comment by christian-bromann** [comment]
> So I walked through all of the options and found the bad ones:

I think we should try to reproduce this with above options, @rpii is there anything else needed for reproduction?

**Comment by rpii** [comment]
So the driver is not custom, just provided via path to avoid io errors on
no Internet systems
. I will look at the options when I return after winter break. However all
problems previously could be replied by config changes in the json

**Comment by christian-bromann** [comment]
Closing due to inactivity.
